### ISSUE TITLE
Event protection and encoding

### ISSUE THREAD

**Issue by martinthomson** [patcg-individual-drafts/ipa#5]
Right now, we're a little imprecise when we talk about how the encryption of various pieces is put in place.

Each match key is protected toward the three helpers.  There is not a lot we can do to optimize there.  Though I will observe that we don't need to explicitly encode a lot of the data.  It is sufficient to include some fields under an AAD.  That trick can provide integrity protection for info that is public, but needs to be bound to the value, like site name and epoch.

Aside from the match key, most of the data comes directly from the report collector.  This information can be protected using bulk transfer keys.  For that, we can use TLS if the report collector talks directly to helpers, or a single large encrypted item otherwise.

That leads to each helper receiving a stream of records that comprise:

1. Site name (cleartext)
2. Epoch (cleartext)
3. An encrypted match key (which need only contain the XOR shares of the match key; the AAD can bind to the site name and epoch)
4. XOR share of the attribution constraint ID (this might be reduced to a share of a single bit, indicating whether to attribute to the previous items if events are presorted)
5. XOR share of the timestamp (this might be removed entirely if items are presorted)
6. A share of a bit indicating if this is a trigger or source event.
7. An additive share of a value.  For a source event, the breakdown key; for a trigger event, the trigger value.

Helpers will use the first three fields to decrypt their shares of the match key, then execute the sort on that field.  There should be no need to sort on attribution constraint ID or timestamp if we require that the items are presorted.

If we want to include multiple breakdowns, we can then do multiple values.  The reason you might want separate fields for breakdown key and trigger value is that they might have different ranges.  However, I think it would be enough to bound the range of values by construction (i.e., share on $\mathbb{Z}_{2^n}$ and assume a contribution of $2^n-1$ - or even $2^n$[^1] - always), but pass the number of breakdown keys for each breakdown up front.  If the breakdown keys exceed the limit, there is no privacy consequence because those items with bad breakdown keys contribute nothing.

Something worth noting here is that there will be a lot of repetitions for the public information.  Site name and epoch will repeat often.  It might be best for encoding reasons to assign these (either individually or as a tuple) a small integer value so that they can be sent without spending too many bytes.  Note that trigger events on a trigger query will have the same site name on every item; same for source events on a source query - this information can therefore be encoded on zero bits.

For anything less than 127 different sites and epochs, the public data might encode in one byte, and two bits should cover all but the most extreme uses.  Then with 48 bytes of HPKE (x25519 and AES-GCM or ChaCha20Poly1305) overhead we have 54 bytes for a 64-bit match key.  That's the big part done.  The rest could probably easily fit into one or two bytes per breakdown.  If you have small breakdown values, you might get as many as 8 breakdowns and the site+epoch identifier into 10 bytes, with a total size of 64 bytes per item.

Packing values this tightly will mean that we need to do a modulus conversion for values, which adds a multiplication round or two.  We can probably do that in parallel with other operations though.  The advantage that comes from doing that is that range checking comes for free, which saves a lot of communication between the report collector and helpers[^2].  

[^1]: The cost of doing $N$ breakdowns with values in the range $[0, M)$ is that you need to ensure that $M\ge N$, which might just mean scaling values up.  That's a little awkward if your range tops out at $2^n-1$, so we might consider adding 1 to the value to make this easier.  If you have 10 breakdowns and you want values from 1 to 4, then multiplying by four to get to a maximum of 16 is easier than working out how to scale up to 15.

[^2]: Prio's range check involves a fairly large number of field elements being submitted.  For a value $[0,2^16)$, this involves sending 20 field elements of something in the order of 80 bits each, so 200 bytes.  That's ~100x more to save the communication of a multiplication.

**Comment by dominiccooney** [comment]
In general, I support these suggestions.

This is a nit, but HPKE [RFC9180 §8.1](https://www.rfc-editor.org/rfc/rfc9180.html#name-auxiliary-authenticated-app) says:

> Applications that only use the single-shot APIs described in [Section 6](https://www.rfc-editor.org/rfc/rfc9180.html#single-shot-apis) should use the Setup info parameter for specifying auxiliary authenticated information.

Since each UA generated report is "one shot" to each helper, presumably we should use info and not aad for the report collector and site name.

**Comment by dominiccooney** [comment]
> Something worth noting here is that there will be a lot of repetitions for the public information. Site name and epoch will repeat often. It might be best for encoding reasons to assign these (either individually or as a tuple) a small integer value so that they can be sent without spending too many bytes.

Or you could specify a compressor like Brotli or zstd and design for that. For example, if you transferred reports as a "struct of arrays" instead of "array of structs" then a compressor will make repeated site names much cheaper than one byte per name. Designing for streaming compression if these are streamed. etc. You will probably get a lot of size wins more easily by dividing the problem that way.

**Comment by martinthomson** [comment]
Good point about info vs. AAD.  I don't know what the efficiency story looks like (HKDF might not be as fast as GHASH for instance), but if this makes the interface easier, the cost difference might be justified.

As for encoding, I believe that a good encoding design can be more efficient than a good compressor, if only for CPU savings.  That comes from the application of domain knowledge.
